stan/math/laplace/laplace_likelihood.hpp

```diff
@@ -0,0 +1,108 @@
+#ifndef STAN_MATH_LAPLACE_LAPLACE_LIKELIHOOD_HPP
+#define STAN_MATH_LAPLACE_LAPLACE_LIKELIHOOD_HPP
+
+#include <stan/math/prim/scal/fun/lgamma.hpp>
+
+namespace stan {
+namespace math {
+
+// TO DO: create a parent structure, with each likelihood
+// function acting as a child structure.
+struct diff_poisson_log {
+  Eigen::VectorXd n_samples_;
+  Eigen::VectorXd sums_;
+
+  diff_poisson_log();
+
+  diff_poisson_log(const Eigen::VectorXd& n_samples,
+                   const Eigen::VectorXd& sums)
+    : n_samples_(n_samples), sums_(sums) { }
+
+  template <typename T>
+  T log_likelihood (const Eigen::Matrix<T, Eigen::Dynamic, 1>& theta)
+    const {
+    double factorial_term = 0;
+    for (int i = 0; i < sums_.size(); i++)
+      factorial_term += lgamma(sums_(i) + 1);
+
+    return - factorial_term
+      + dot_product(theta, sums_)
+      - dot_product(exp(theta), n_samples_);
+  }
+
+  template <typename T>
+  void diff (const Eigen::Matrix<T, Eigen::Dynamic, 1>& theta,
+             Eigen::Matrix<T, Eigen::Dynamic, 1>& gradient,
+             Eigen::Matrix<T, Eigen::Dynamic, 1>& hessian) const {
+    hessian = - elt_multiply(n_samples_, exp(theta));
+    gradient = sums_ + hessian;
+  }
+
+  template <typename T>
+  Eigen::Matrix<T, Eigen::Dynamic, 1>
+  third_diff(const Eigen::Matrix<T, Eigen::Dynamic, 1>& theta) const {
```

Comment:
It's the third derivative tensor. It doesn't have a name, like gradient for first order and Jacobian for second order. This returns a vector because for latent gaussian models, the derivatives tensors are diagonal. I can share the math.